examples/celeba_glow.py

```diff
@@ -0,0 +1,612 @@
+#! -*- coding: utf-8 -*-
+# Keras implement of Glow
+# https://blog.openai.com/glow/
+
+from keras.layers import *
+from keras.models import Model
+from keras.datasets import cifar10
+from keras import backend as K
+from keras.callbacks import Callback
+from keras.optimizers import Adam
+import imageio
+import numpy as np
+from scipy import misc
+import glob
+import os
+
+if not os.path.exists('samples'):
+    os.mkdir('samples')
+
+imgs = glob.glob('img_align_celeba/*.jpg')
+
+height, width = misc.imread(imgs[0]).shape[:2]
+center_height = int((height - width) / 2)
+
+img_size = 64  # for a fast try, please use img_size=32
+depth = 10  # orginal paper use depth=32
+level = 3  # orginal paper use level for 256*256 CelebA HQ
+
+
+def imread(f):
+    x = misc.imread(f)
+    x = x[center_height:center_height + width, :]
+    x = misc.imresize(x, (img_size, img_size))
+    return x.astype(np.float32) / 256 - 0.5
+
+
+def data_generator(batch_size=32):
+    X = []
+    while True:
+        np.random.shuffle(imgs)
+        for f in imgs:
+            X.append(imread(f))
+            if len(X) == batch_size:
+                X = np.array(X)
+                yield X, X.reshape((X.shape[0], -1))
+                X = []
+
+
+class Permute(Layer):
+    """New Permute layer. Reverse or shuffle the final axis of inputs
+    """
+
+    def __init__(self, mode='reverse', **kwargs):
+        super(Permute, self).__init__(**kwargs)
+        self.idxs = None
+        self.mode = mode
+
+    def build(self, input_shape):
+        in_dim = input_shape[-1]
+        if self.idxs is None:
+            if self.mode == 'reverse':
+                self.idxs = self.add_weight(
```

Comment:
These are not weights that are updated during training, thus you do not need to create them with `add_weight`. You could simply create tensors and set them as layer attributes (make sure the random state is seeded so you get the same weights when loading a saved model).

Reply:
I don't agree this. self.idxs is actually a (non-trainable but random) weight for this layer, so I use `add_weight` to ask model save this weights. I think it is a standard usage of keras custom layer.